File: android/app/src/main/java/com/plantora/billing/ui/sales/ApprovalsViewModel.kt
package com.plantora.billing.ui.sales

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.plantora.billing.data.SettlementRepository
import com.plantora.billing.data.remote.friendlyError
import com.plantora.billing.domain.PendingSettlement
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * The manager's queue of salesperson due-collections awaiting approval. Approving
 * applies the cash/UPI split to the bill and closes the due; rejecting leaves the
 * due outstanding.
 */
data class ApprovalsUiState(
    val loading: Boolean = true,
    val error: String? = null,
    val items: List<PendingSettlement> = emptyList(),
    val actingId: String? = null,
    val message: String? = null,
)

@HiltViewModel
class ApprovalsViewModel @Inject constructor(
    private val repo: SettlementRepository,
) : ViewModel() {

    private val _ui = MutableStateFlow(ApprovalsUiState())
    val ui: StateFlow<ApprovalsUiState> = _ui.asStateFlow()

    init { load() }

    fun load() {
        _ui.update { it.copy(loading = true, error = null) }
        viewModelScope.launch {
            runCatching { repo.pending() }
                .onSuccess { list -> _ui.update { it.copy(loading = false, items = list) } }
                .onFailure { e -> _ui.update { it.copy(loading = false, error = friendlyError(e)) } }
        }
    }

    fun approve(item: PendingSettlement) = act(item, approve = true)
    fun reject(item: PendingSettlement) = act(item, approve = false)

    private fun act(item: PendingSettlement, approve: Boolean) {
        if (_ui.value.actingId != null) return
        _ui.update { it.copy(actingId = item.id) }
        viewModelScope.launch {
            runCatching { if (approve) repo.approve(item.id) else repo.reject(item.id) }
                .onSuccess {
                    _ui.update { s ->
                        s.copy(
                            actingId = null,
                            items = s.items.filterNot { it.id == item.id },
                            message = if (approve) "Approved — ${item.amount.format()} collected." else "Rejected.",
                        )
                    }
                }
                .onFailure { e -> _ui.update { it.copy(actingId = null, message = friendlyError(e, "Couldn't update the request.")) } }
        }
    }

    fun dismissMessage() = _ui.update { it.copy(message = null) }
}

File: Plantbill/Plantbill/Features/Sales/ApprovalsViewModel.swift
import Foundation
import Observation

/// The manager's queue of salesperson due-collections awaiting approval.
/// Approving applies the cash/UPI split to the bill and closes the due;
/// rejecting leaves the due outstanding.
@Observable
@MainActor
final class ApprovalsViewModel {
    private(set) var isLoading = true
    private(set) var loadError: String?
    private(set) var items: [PendingSettlement] = []
    private(set) var actingId: UUID?
    var message: String?

    func load() async {
        isLoading = true
        loadError = nil
        do {
            let list: PendingSettlementList = try await APIClient.shared.send(Endpoint(path: "settlements/pending"))
            items = list.items
            isLoading = false
        } catch let error as APIError {
            isLoading = false
            loadError = error.userMessage
        } catch {
            isLoading = false
            loadError = APIError.unknown.userMessage
        }
    }

    func approve(_ item: PendingSettlement) async { await act(item, approve: true) }
    func reject(_ item: PendingSettlement) async { await act(item, approve: false) }

    private func act(_ item: PendingSettlement, approve: Bool) async {
        guard actingId == nil else { return }
        actingId = item.id
        do {
            let path = "settlements/\(item.id)/\(approve ? "approve" : "reject")"
            let _: SettlementActionResult = try await APIClient.shared.send(Endpoint(path: path, method: .post))
            items.removeAll { $0.id == item.id }
            message = approve ? "Approved — \(item.amountMoney.format()) collected." : "Rejected."
            actingId = nil
        } catch let error as APIError {
            actingId = nil
            message = error.userMessage
        } catch {
            actingId = nil
            message = APIError.unknown.userMessage
        }
    }

    func dismissMessage() { message = nil }
}
